sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,829 @@
+                        data_len: 4u64,
+                        is_negative: false,
+                    }
+                };
+                unsafe {
+                    sol_bignum_from_u32(
+                        &mut bn_syscall_result as *mut _ as *mut u64,
+                        u32_val as u64,
+                    );
+                    value.set_len(bn_syscall_result.data_len as usize);
+                }
+                Self {
+                    is_negative: bn_syscall_result.is_negative,
+                    data: value,
+                }
+            }
+        }
+    }
+
+    /// Returns a BigNumber from a decimal string
+    pub fn from_dec_str(string: &str) -> Self {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            let bn = openssl::bn::BigNum::from_dec_str(string).unwrap();
+            Self { data: bn }
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            extern "C" {
+                fn sol_bignum_from_dec_str(
+                    in_dec_str_ptr: *const u64,
+                    in_size: u64,
+                    bn_syscall_result: *mut u64,
+                ) -> u64;
+            }
+            // The output buffer could be sized to 3.32 bits/digit but this would consume compute units
+            let mut value = Vec::<u8>::with_capacity(string.len());
+            let mut bn_syscall_result = {
+                FfiBigNumber {
+                    data: value.as_mut_ptr() as *mut _ as u64,
+                    data_len: string.len() as u64,
+                    is_negative: false,
+                }
+            };
+
+            unsafe {
+                sol_bignum_from_dec_str(
+                    string.as_ptr() as *const _ as *const u64,
+                    string.len() as u64,
+                    &mut bn_syscall_result as *mut _ as *mut u64,
+                );
+                value.set_len(bn_syscall_result.data_len as usize);
+            }
+            Self {
+                is_negative: bn_syscall_result.is_negative,
+                data: value,
+            }
+        }
+    }
+
+    /// Construct a BigNumber from a unsigned absolute big endian array of bytes
+    pub fn from_slice(in_data: &[u8]) -> Self {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            Self {
+                data: openssl::bn::BigNum::from_slice(in_data).unwrap(),
+            }
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            extern "C" {
+                fn sol_bignum_from_slice(
+                    rhs_ffi_in_addr: *const u64,
+                    bn_syscall_result: *mut u64,
+                ) -> u64;
+            }
+            let data_size = in_data.len();
+            let rhs_ffi_in = {
+                FfiBigNumber {
+                    data: in_data.as_ptr() as *const _ as u64,
+                    data_len: data_size as u64,
+                    is_negative: false,
+                }
+            };
```

Comment:
What's the difference between `rhs_ffi_in` and the result returned? Is this not the same?

Reply:
Apart from the BigNum struct of course, so yo could  just return this:
```
Self {
   data: in_data.to_vec(),
   is_negative: false,
}
```

Reply:
It should be the same assuming it passes the bounds check in the syscall. As noted above, the roundtrip does the bounds check and data costing. Also has to be available to "C"

Reply:
"C" doesn't need this, this is trivial to implement in C.